README.md
# UizaSDK
<!-- [![CI Status](http://img.shields.io/travis/uizaio/UizaSDK.svg?style=flat)](https://travis-ci.org/uizaio/UizaSDK) -->
[![Version](https://img.shields.io/cocoapods/v/UizaSDK.svg?style=flat)](http://cocoapods.org/pods/UizaSDK)
[![License](https://img.shields.io/cocoapods/l/UizaSDK.svg?style=flat)](http://cocoapods.org/pods/UizaSDK)
[![Platform](https://img.shields.io/cocoapods/p/UizaSDK.svg?style=flat)](http://cocoapods.org/pods/UizaSDK)
![Swift](https://img.shields.io/badge/%20in-swift%205.0-orange.svg)

(Kéo xuống phía dưới để đọc tiếng Việt)

UizaSDK is a framework to connect to Uiza system

## Compatibility

UizaSDK requires Swift 4.0+ and iOS 8+

## Installation


### CocoaPods

To integrate UizaSDK into your Xcode project using [CocoaPods](http://cocoapods.org), specify it in your `Podfile`:

```ruby
pod 'UizaSDK'
```

Then run the following command:

```bash
$ pod install
```

## Usage

## Framework Init
Always initialize the framework by the following line before calling any API functions:

``` swift
import UizaSDK

UizaSDK.initWith(appId: YOUR_APP_ID, token: TOKEN, api: YOUR_DOMAIN, version: API_VERSION)
```

YOUR_APP_ID and YOUR_DOMAIN : get from registration email

TOKEN: generate from https://docs.uiza.io/#get-api-key

API_VERSION: version of target API, set .v3 or .v4 (default is .v3)

## Call API
``` swift
UZContentServices().loadDetail(entityId: ENTITY_ID, completionBlock: { (videoItem, error) in
  if error != nil {
    print("Error: \(error)")
  }
  else {
    print("Video: \(videoItem)")
  }
})
```

## How to play video
``` swift
let playerViewController = UZPlayerViewController()
playerViewController.player.loadVideo(entityId: ENTITY_ID)
present(playerViewController, animated: true, completion: nil)
```

 You might have to add these lines to `Info.plist` to disable App Transport Security (ATS) to be able to play video:
``` xml
<key>NSAppTransportSecurity</key>  
<dict>  
  <key>NSAllowsArbitraryLoads</key><true/>  
</dict>
```

## How to broadcast livestream
``` swift
let viewController = UZLiveStreamViewController()
viewController.liveEventId = ENTITY_ID
self.present(viewController, animated: true, completion: nil)
```

Remember to add these usage description keys into `Info.plist` file:
``` xml
<key>NSCameraUsageDescription</key>
<string>App needs access to camera for livestream</string>
<key>NSMicrophoneUsageDescription</key>
<string>App needs access to microphone for livestream</string>
```

## Change Player Themes
``` swift
let playerViewController = UZPlayerViewController()
playerViewController.player.theme = UZTheme1()
```

UizaPlayer currently has 7 built-in themes:

[UZTheme1](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme1.jpg)

[UZTheme2](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme2.jpg)

[UZTheme3](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme3.jpg)

[UZTheme4](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme4.jpg)

[UZTheme5](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme5.jpg)

[UZTheme6](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme6.jpg)

[UZTheme7](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme7.jpg)

## Create CustomTheme

You can create your own custom theme by creating a class inheriting from [UZPlayerTheme Protocol](https://uizaio.github.io/uiza-sdk-player-ios/Protocols/UZPlayerTheme.html) following this template: [UZCustomTheme](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/UZCustomTheme.swift)

You can also create your custom end screen by subclassing `UZEndscreenView`, then set an instance to `player.controlView.endscreenView`
``` swift
self.playerViewController.player.controlView.endscreenView = MyCustomEndScreen()
```

## Create Player with Floating Mode

You can create player with "drag down to floating mode" like Facebook or Youtube, by subclassing [UZFloatingPlayerViewController](https://uizaio.github.io/uiza-sdk-player-ios/Classes/UZFloatingPlayerViewController.html), then you can add more UI for displaying video details and add them to  `detailsContainerView` 

Then present using this code:
``` swift
UZFloatingPlayerViewController().present(with: videoItem, playlist: playlist)
```

See [Example](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/Example/UizaSDKExample/FloatingPlayerViewController.swift)

For API details, check [API Document](https://uizaio.github.io/uiza-sdk-player-ios/)

## Google ChromeCast supports
If developing using Xcode 10 and targeting iOS devices running iOS 12 or higher, the "Access WiFi Information" capability is required in order to discover and connect to Cast devices
![](https://developers.google.com/cast/images/xcode_wifi_capability_error.png)

## Support
namnh@uiza.io

## License

UizaSDK is released under the BSD license. See [LICENSE](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/LICENSE) for details.

----------------------------------------------------------------

# UizaSDK

UizaSDK là bộ Framework hỗ trợ kết nối đến API của hệ thống Uiza

## Tương Thích

UizaSDK yêu cầu Swift 4.0_ và iOS 8+

## Cài Đặt

### CocoaPods

Cài đặt thông qua [CocoaPods](http://cocoapods.org)

Thêm vào `Podfile` dòng sau:

```ruby
pod 'UizaSDK'
```

Sau đó chạy lệnh này:

```bash
$ pod install
```

## Cách Sử Dụng

## Khởi tạo
Luôn khởi động framework này trước khi gọi bất cứ hàm API nào bằng cách gọi lệnh sau:

``` swift
import UizaSDK

UizaSDK.initWith(appId: YOUR_APP_ID, token: TOKEN, api: YOUR_DOMAIN, version: API_VERSION)
```

YOUR_APP_ID và YOUR_DOMAIN : lấy từ thông tin trong email đăng ký

TOKEN: được tạo từ trang https://docs.uiza.io/#get-api-key

API_VERSION: phiên bản API sử dụng, truyền vào .v3 hoặc .v4 (nếu không truyền, mặc định là .v3)

## Gọi hàm API
``` swift
UZContentServices().loadDetail(entityId: ENTITY_ID, completionBlock: { (videoItem, error) in
  if error != nil {
    print("Error: \(error)")
  }
  else {
    print("Video: \(videoItem)")
  }
})
```

## Cách play video
``` swift
let playerViewController = UZPlayerViewController()
playerViewController.player.loadVideo(entityId: ENTITY_ID)
present(playerViewController, animated: true, completion: nil)
```

Nếu gặp trường hợp video không play được do vấn đề App Transport Security (ATS), bạn phải thêm dòng sau vào file `Info.plist` để có thể play được video:
``` xml
<key>NSAppTransportSecurity</key>  
<dict>  
  <key>NSAllowsArbitraryLoads</key><true/>  
</dict>
```

## Cách phát livestream
``` swift
let viewController = UZLiveStreamViewController()
viewController.liveEventId = [ENTITY_ID]
self.present(viewController, animated: true, completion: nil)
```

Nhớ thêm mô tả yêu cầu chức năng truy cập camera và micro vào file `Info.plist`:
``` xml
<key>NSCameraUsageDescription</key>
<string>App cần truy cập camera để phát livestream</string>
<key>NSMicrophoneUsageDescription</key>
<string>App cần truy cập micro để thu tiếng khi phát livestream</string>
```

## Thay đổi giao diện
``` swift
let playerViewController = UZPlayerViewController()
playerViewController.player.theme = UZTheme1()
```

UizaPlayer có sẵn 7 giao diện sau:

[UZTheme1](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme1.jpg)

[UZTheme2](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme2.jpg)

[UZTheme3](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme3.jpg)

[UZTheme4](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme4.jpg)

[UZTheme5](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme5.jpg)

[UZTheme6](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme6.jpg)

[UZTheme7](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/theme7.jpg)

## Tự tạo giao diện (CustomTheme)

Bạn có thể tự tạo giao diện riêng bằng cách tạo class kế thừa [UZPlayerTheme Protocol](https://uizaio.github.io/uiza-sdk-player-ios/Protocols/UZPlayerTheme.html) theo mẫu code này: [UZCustomTheme](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/themes/UZCustomTheme.swift)

Bạn cũng có thể thay đổi giao diện của màn hình kết thúc bằng cách tạo class kế thừa `UZEndscreenView`, sau đó đưa nó vào `player.controlView.endscreenView`
``` swift
self.playerViewController.player.controlView.endscreenView = MyCustomEndScreen()
```

## Tạo Player với Floating Mode

Bạn có thể tạo player với chức năng "kéo xuống góc màn hình" giống Facebook hoặc Youtube bằng cách tạo class kế thừa [UZFloatingPlayerViewController](https://uizaio.github.io/uiza-sdk-player-ios/Classes//UZFloatingPlayerViewController.html), trong đó bạn có thể thêm các UI hiển thị chi tiết của video và đưa vào `detailsContainerView` 

Sau đó present bằng lệnh sau:
``` swift
UZFloatingPlayerViewController().present(with: videoItem, playlist: playlist)
```

Xem [Ví dụ](https://github.com/uizaio/uiza-sdk-player-ios/blob/master/Example/UizaSDKExample/FloatingPlayerViewController.swift)

## Hỗ trợ Google ChromeCast
Nếu bạn sử dụng Xcode 10 và phát triển cho iOS 12 trở lên, cần phải bật "Access WiFi Information" thì chức năng tìm kiếm các thiết bị Cast mới có thể hoạt động
![](https://developers.google.com/cast/images/xcode_wifi_capability_error.png)

Xem chi tiết [Tài liệu API](https://uizaio.github.io/uiza-sdk-player-ios/)

## Hỗ Trợ
namnh@uiza.io


CHANGELOG.md
### Version 8.4.12 (Mar 6 2020)
- [Fixed] UZPlayer did not recognize live video when loadVideo with entityID

### Version 8.4.7 (Dec 23 2019)
- [Added] Able to check status of a live feed:
``` swift
UZLiveService().checkSignalStatus(entityId: "")
```

### Version 8.3.7 (Aug 6 2019)
- [Added] Set adaptiveBitrate (UZLiveStreamViewController) to auto change bitrate base on network performance
- [Updated] Updated video configuration & audio configuration for livestream
- [Updated] Fullscreen in portrait if video is portrait

### Version 8.3 (Jul 10 2019)
- [Updated] Supports FullHD 1080 & UltraHD 4K for livestreamming
- [Updated] Supports external subtitle
- [Updated] Auto retry when livestream stalled

### Version 8.2.3 (Jul 10 2019)
- [Updated] Auto check for live latency. You can set max latency interval by setting `maxLatencyTime` in UZPlayer

### Version 8.2.1 (Jul 10 2019)
- [Updated] Improved app inactive/active event handling
- [Added] Support loading subtitles

### Version 8.2 (Jul 10 2019)
- [Fixed] Crash bug fixes
- [Updated] UZPlayer updated, please note that we have renammed these protocol functions:
From:
``` swift
	func UZPlayer(player: UZPlayer, playerStateDidChange state: UZPlayerState)
	func UZPlayer(player: UZPlayer, loadedTimeDidChange loadedDuration: TimeInterval, totalDuration: TimeInterval)
	func UZPlayer(player: UZPlayer, playTimeDidChange currentTime : TimeInterval, totalTime: TimeInterval)
	func UZPlayer(player: UZPlayer, playerIsPlaying playing: Bool)
}
```
To:
``` swift
	func player(player: UZPlayer, playerStateDidChange state: UZPlayerState)
	func player(player: UZPlayer, loadedTimeDidChange loadedDuration: TimeInterval, totalDuration: TimeInterval)
	func player(player: UZPlayer, playTimeDidChange currentTime : TimeInterval, totalTime: TimeInterval)
	func player(player: UZPlayer, playerIsPlaying playing: Bool)
}
```

### Version 8.1.2 (Jul 09 2019)
- [Added] Measure network latency in visualize info

### Version 8.1 (Jun 26 2019)
- [Updated] Supports mixed Swift versions

### Version 8.0 (Jun 26 2019)
- [New] UizaSDK is now released as Open Source

### Version 7.6.2 (Jun 12 2019)
- [Added] New Visualize information for debugging

### Version 7.6 (Jun 12 2019)
- [Updated] Google Ads and Google Cast are now optional

### Version 7.2 (Apr 27 2019)
- [Updated] Added Sentry error log

### Version 7.1 (Apr 27 2019)
- [Updated] Able to minimize player to bottom-left or bottom-right corner base on user's demand

### Version 7.0 (Mar 29 2019)
- [Updated] supports both API v3 and API v4
- [Fixed+Updated] able to set interval to auto stop livestream when closing app to background

### Version 6.8.8 (Jan 15 2019)
- [Fixed] unable to set livestream camera position on initialization
- [Fixed] duplicate function name might cause build error

### Version 6.8.7 (Jan 09 2019)
- [Fixed] fix audioBitrate metadata push in livestream library

### Version 6.8.6 (Jan 04 2019)
- [Fixed] unable to drag-to-minimize player after showing an alert

### Version 6.8.5 (Jan 03 2019)
- [Updated] player will keep pausing when back from background state. For auto resume, set this: 
``` swift
player.autoResumeWhenBackFromBackground = true
```

### Version 6.8.4 (Dec 28 2018)
- [Updated] able to disable/enable floating mode by setting: 
``` swift
floatingPlayerViewController.floatingHandler.isEnabled = true/false
```

### Version 6.8.3 (Dec 26 2018)
- [Updated] change message when live ended directly with 
``` swift
playerViewController.liveEndedMessage
```

### Version 6.8.2 (Dec 21 2018)
- [Updated] show message when live video has ended.

### Version 6.8 (Dec 18 2018)
- [New] use pod 'UizaSDK_8' to install SDK for project with deployment target from 8.x
Please note that this SDK does not support:
- Google Cast
- Google Ads IMA
- Picture in Picture
- Buffer duration adjustment
- Future updates like DRM, video download etc..

### Version 6.7 (Dec 17 2018)
- [Fix] Set custom LiveStreamUIView did not catch button events
- [Update] Downgrade minumum development target to iOS 9.x

### Version 6.6.5 (Dec 12 2018)
- [Update] more supports for customizing UZLiveStreamViewController, now you can customize videoConfiguration and audioConfiguration (see example)

### Version 6.6.4 (Dec 04 2018)
- [Update] able to drag the floating video out of screen to dismiss
- [Fix] UZFloatingPlayerViewController did not dismiss properly.

### Version 6.6.1 (Nov 30 2018)
- [Update] toggle timeshift for live video:
``` swift
playerController.controlView.enableTimeShiftForLiveVideo = true/false
```

### Version 6.6 (Nov 26 2018)
- [Update] Able to dock floating video to screen corners
``` swift
floatingPlayerViewController.floatingHandler.allowsCornerDocking = true
```

### Version 6.5.3 (Nov 23 2018)
- [Fix] Invalid logging URL
- [Update] Timeshift for live video
- [Update] Better support for setting custom playerViewController for UZFloatingPlayerViewController

### Version 6.5 (Oct 25 2018)
- [New] UZFloatingPlayerViewController to present player with drag down to floating mode gesture. 

### Version 6.4.4 (Oct 25 2018)
- [Fixed] Show endscreen unexpectedly when seeking after the player has ended a movie

### Version 6.4.2 (Oct 14 2018)
- [Fixed] IMA Ads did not work
- [Fixed] Player config loading not work

### Version 6.4 (Oct 14 2018)
- [New] Supports IMA Ads 

### Version 6.3.9 (Oct 08 2018)
- [Updated] New linkplay for Staging enviroment 

### Version 6.3.8 (Oct 04 2018)
- [Fixed] Unable to change progress color for custom time slider 

### Version 6.3.7 (Oct 01 2018)
- More supports for overriding UZPlayer

### Version 6.3.6 (Sep 27 2018)
- Supports overriding UZPlayer

### Version 6.3 (Sep 23 2018)

- Supports XCode 10, Swift 4.2
- New UZThemeConfig to fetch player config from workspace
``` swift
player.loadConfigId(configId: [CONFIG_ID])
```


UizaSDK/SDK/Models/UZUser.swift
//
//  UZUser.swift
//  UizaSDK
//
//  Created by Nam Kennic on 4/26/17.
//  Copyright © 2017 Nam Kennic. All rights reserved.
//

import UIKit

// swiftlint:disable identifier_name
let CURRENT_USER_KEY = "com.uiza.currentUser.3_0"

/**
Info of user
*/
open class UZUser: UZModelObject {
	/** `id` of user */
	public var id: String!
	/** Name of user */
	public var name: String?
	/** Gender of user */
	public var gender: String? // male / female
	/** Mobile phone */
	public var mobile: String?
	/** Email address */
	public var email: String?
	/** Birthday of user */
	public var birthday: Date?
	/** Facebook ID, if logged in from Facebook */
	public var facebookId: String?
	/** Avatar URL */
	public var avatarURL: URL?
	/** Token of user */
	public var token: String?
	
	/** `true` if this is you */
    public var isMe: Bool {
        if let currentUser = UZUser.currentUser {
            return self.id == currentUser.id
        }
        
        return false
    }
	
	fileprivate static var _currentUser: UZUser?
	/** Logged in user, returns `nil` if user is not logged in */
	public class var currentUser: UZUser? {
        if _currentUser == nil {
            let userData: Data? = UserDefaults.standard.object(forKey: CURRENT_USER_KEY) as? Data
            if userData != nil {
                _currentUser = NSKeyedUnarchiver.unarchiveObject(with: userData!) as? UZUser
            }
        }
        
        return _currentUser
		/*
		set (newValue) {
			if (_currentUser != newValue) {
				_currentUser = newValue
				
				if _currentUser != nil {
					let userData: Data = NSKeyedArchiver.archivedData(withRootObject: _currentUser!)
					UserDefaults.standard.set(userData, forKey: CURRENT_USER_KEY)
					UserDefaults.standard.synchronize()
				}
				else {
					UserDefaults.standard.removeObject(forKey: CURRENT_USER_KEY)
					UserDefaults.standard.synchronize()
				}
			}
		}
		*/
	}
	
//	/** Token của Facebook nếu đây là người dùng được login bằng Facebook */
//	public var facebookToken : FBSDKAccessToken? {
//		didSet {
//			if facebookToken != nil {
//				data?.setValue(facebookToken, forKey:"facebookToken")
//			}
//			else {
//				data?.removeObject(forKey: "facebookToken")
//			}
//		}
//	}
//
//	/** Facebook Profile nếu đây là người dùng được login bằng Facebook */
//	public var facebookProfile : FBSDKProfile? {
//		didSet {
//			if facebookProfile != nil {
//				data?.setValue(facebookProfile, forKey:"facebookProfile")
//			}
//			else {
//				data?.removeObject(forKey: "facebookProfile")
//			}
//		}
//	}
	
	/**
	URL of Facebook avatar if user has logged in from Facebook
	- parameter size: Kích thước hình đại diện
	*/
	public func facebookAvatar(size: Int = 100) -> URL? {
		if facebookId != nil && !facebookId!.isEmpty {
			let finalSize = size * Int(UIScreen.main.scale)
			let result = "https://graph.facebook.com/\(facebookId!)/picture?width=\(finalSize)"
			return URL(string: result)
		}
		
		return nil
	}
	
	// MARK: -
	
	internal func saveAsCurrentUser() {
		UZUser._currentUser = self
//		self.facebookToken = FBSDKAccessToken.current()
		
		let userData: Data = NSKeyedArchiver.archivedData(withRootObject: self)
		UserDefaults.standard.set(userData, forKey: CURRENT_USER_KEY)
		UserDefaults.standard.synchronize()
	}
	
	internal class func clearCurrentUser() {
		UZUser._currentUser = nil
//		FBSDKAccessToken.setCurrent(nil)
//		FBSDKProfile.setCurrent(nil)
		
		UserDefaults.standard.removeObject(forKey: CURRENT_USER_KEY)
		UserDefaults.standard.synchronize()
	}
	
	func merge(with user: UZUser!) {
		if self.data == nil {
			self.data = NSMutableDictionary()
		}
		
		let lastToken: String = String(describing: self.token ?? "") // copy to new instance
        self.data!.addEntries(from: user.data! as? [String: AnyHashable] ?? [:])
		self.data!.setObject(lastToken, forKey: "token" as NSCopying)
		parse(user.data)
		
		self.token = lastToken
	}
	
	// MARK: -
	
	public convenience init(data: NSDictionary) {
		self.init()
		
		self.data = data.mutableCopy() as? NSMutableDictionary
		self.parse(data)
	}
	
	public override init() {
		super.init()
	}
	
	required public init(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
		
		self.data = aDecoder.decodeObject(forKey: "data") as? NSMutableDictionary
		self.parse(self.data)
	}
	
	override open func encode(with aCoder: NSCoder) {
		if self.data != nil {
			aCoder.encode(self.data, forKey: "data")
		}
	}
	
	override func parse(_ data: NSDictionary?) {
		if data != nil {
			let userData 	= data!.value(for: "user", defaultValue: nil) as? NSDictionary ?? data
			
			id				= userData!.string(for: "id", defaultString: "")
			name			= userData!.string(for: "name", defaultString: nil)
			gender			= userData!.string(for: "gender", defaultString: nil)
			email			= userData!.string(for: "email", defaultString: nil)
			mobile			= userData!.string(for: "phone", defaultString: nil)
			token			= userData!.string(for: "token", defaultString: nil)
			facebookId		= userData!.string(for: "fbId", defaultString: nil)
			
//			if let birthdayString = userData!.string(for: "birthday", defaultString: nil) {
//				birthday = birthdayString.length>10 ? Date(fromString: birthdayString, format: .isoDateTimeMilliSec) : Date(fromString: birthdayString, format: .custom("YYYY-MM-dd"))
//			}
			
//			if let fbToken = data!["facebookToken"] as? FBSDKAccessToken {
//				self.facebookToken = fbToken
//			}
//
//			if let fbProfile = data!["facebookProfile"] as? FBSDKProfile {
//				self.facebookProfile = fbProfile
//			}
			
			if token == nil {
				token = data!.string(for: "token", defaultString: nil)
			}

		}
	}
	
	/** Object description */
	override open var description: String {
		return "\(super.description) [\(name ?? "")] [token:\(token ?? "")]"
	}
	
	open override func isEqual(_ object: Any?) -> Bool {
		if object is UZUser {
            return (object as? UZUser)?.id == self.id
		}
		
		return false
	}
	
	static public func == (lhs: UZUser, rhs: UZUser) -> Bool {
		return lhs.id == rhs.id
	}
	
}
